File: 224.Basic-Calculator.java
// https://leetcode.com/problems/basic-calculator/
//
// algorithms
// Hard (33.01%)
// Total Accepted:    110,510
// Total Submissions: 334,771

class Solution {
    public int calculate(String s) {
        LinkedList<Integer> stack = new LinkedList<>();
        
        char[] chArr = s.toCharArray();
        int len = chArr.length, i = 0;
        int flag = 1;
        boolean firstNum = true;
        int cur = 0;
        
        while (i < len) {
            if (chArr[i] == ' ') {
                i++;
                continue;
            } else if (chArr[i] == '+') {
                flag = 1;
            } else if (chArr[i] == '-') {
                flag = -1;
            } else if (chArr[i] == '(') {
                stack.offerLast(cur);
                stack.offerLast(flag);
                firstNum = true;
            } else if (chArr[i] == ')') {
                flag = stack.pollLast();
                cur = stack.pollLast() + (flag * cur);
            } else {
                int j = i + 1;
                while (j < len && Character.isDigit(chArr[j])) {
                    j++;
                }
                int num = Integer.valueOf(s.substring(i, j));
                if (firstNum) {
                    firstNum = false;
                    cur = num;
                } else {
                    cur += flag * num;
                }
                i = j - 1;
            }
            i++;
        }
        
        return cur;
    }
}

File: 224.Basic-Calculator.py
# https://leetcode.com/problems/basic-calculator/
#
# algorithms
# Hard (33.82%)
# Total Accepted:    120,177
# Total Submissions: 355,380
# beats 65.45% of python submissions


from collections import deque


class Solution(object):
    def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        s = s.strip()
        length = len(s)

        stack = deque()
        cur_num, cur_sample = 0, 1

        idx = 0

        while idx < length:
            ch = s[idx]
            if ch == ' ':
                idx += 1
                continue

            if ch.isdigit():
                j = idx + 1
                while j < length and s[j].isdigit():
                    j += 1

                num = int(s[idx:j])
                cur_num += (cur_sample * num)
                idx = j
                continue

            if ch == '+':
                cur_sample = 1

            elif ch == '-':
                cur_sample = -1

            elif ch == '(':
                stack.append((cur_num, cur_sample))
                cur_num, cur_sample = 0, 1

            else:
                pre_num, pre_sample = stack.pop()
                cur_sample = 1
                cur_num = pre_num + pre_sample * cur_num

            idx += 1

        return cur_num
